src/dtls_connection_api.c

```diff
@@ -446,159 +519,165 @@ static int dtls_connection_start(void *context_p)
-                {
-                    dtls_connection_close(context, conn);
-                }
+    FD_ZERO(&read_fds);
+    FD_SET(sock, &read_fds);
 
-                *connection = conn;
-                return ret;
-            }
-        }
+    ret = select(sock + 1, &read_fds, NULL, NULL, tv);
+    if (ret <= 0)
+    {
+        return ret;
     }
 
-//  expecting a new client connection
     if (FD_ISSET(sock, &read_fds))
     {
         context->conn_listen->addr_size = sizeof(context->conn_listen->addr);
 
-//      only peek here so that if connection is successful, socket is still active when passing to handshake
-        ret = recvfrom(sock, buffer, size, MSG_PEEK,
-                       (struct sockaddr *)&context->conn_listen->addr, &context->conn_listen->addr_size);
+        ret = recvfrom(sock, buffer, size, MSG_PEEK, (struct sockaddr *)&context->conn_listen->addr,
+                       &context->conn_listen->addr_size);
         if (ret > 0)
         {
-            memset(&prestate, 0, sizeof(gnutls_dtls_prestate_st));
+            conn = dtls_connection_find(context->connection_list, &context->conn_listen->addr,
+                                        context->conn_listen->addr_size);
 
-            ret = gnutls_dtls_cookie_verify(&context->cookie_key, &context->conn_listen->addr,
-                                            sizeof(context->conn_listen->addr), buffer, ret, &prestate);
-            if (ret == GNUTLS_E_BAD_COOKIE)
+            if (conn == NULL)
             {
-                gnutls_dtls_cookie_send(&context->cookie_key, &context->conn_listen->addr,
-                                        sizeof(context->conn_listen->addr), &prestate, context->conn_listen, dtls_connection_net_send);
-            }
-            else if (ret == 0)
-            {
-                ret = -1;
-                conn = context->conn_listen;
+                memset(&prestate, 0, sizeof(gnutls_dtls_prestate_st));
+
+                ret = gnutls_dtls_cookie_verify(&context->cookie_key, &context->conn_listen->addr,
+                                                sizeof(context->conn_listen->addr), buffer, ret, &prestate);
 
-                context->conn_listen = dtls_connection_new_listen(context);
-                if (context->conn_listen == NULL)
+                if (ret == GNUTLS_E_BAD_COOKIE)
                 {
-                    context->conn_listen = conn;
-                    goto connect_fail;
+                    gnutls_dtls_cookie_send(&context->cookie_key, &context->conn_listen->addr,
+                                            sizeof(context->conn_listen->addr), &prestate, context->conn_listen, dtls_connection_net_send);
+                    recvfrom(sock, buffer, size, 0, NULL, NULL);
                 }
-
-//              the current socket will be taken over by the client connection
-                if (connect(conn->sock, (struct sockaddr *)&conn->addr, sizeof(conn->addr)))
+                else if (ret == 0)
```

Comment:
Since return code comes from another library it would be better to use it's definition, i.e. `GNUTLS_E_SUCCESS`